rm: linux: Add DEV_MODE description (#577)
foundriesio/docs: source/reference-manual/linux/linux-dev-mode.rst

.. _ref-dev-mode:

Development Mode
=================================

LmP provides a variable ``DEV_MODE`` to be used to enable a development mode
defined by the factory source code.

> The variable is defined as ``lmp:params:DEV_MODE`` and can be set on

Reviewer: I'm not sure a user knows what to set for `DEV_MODE` or where.  Should we have a small snippet of the `factory-config.yml` file?

```
The variable is defined as ``lmp:params:DEV_MODE`` and can be configured in
:ref:`factory-config.yml <def-lmp>` in ``ci-scripts.git`` with::

  lmp:
    params:
      DEV_MODE: 1
```

And we can also add `DEV_MODE` to our Factory Definition file:
https://docs.foundries.io/latest/reference-manual/factory/factory-definition.html

Author: I agree with the first part. But I'm confused about what you mean by "add DEV_MODE to our Factory Definition file".  Do you mean it would be good to have a list of all supported variables there? (I can remember at least 2 others) @mike-scott

Reviewer: I was thinking of documenting the extra variables here:
https://docs.foundries.io/latest/reference-manual/factory/factory-definition.html
Making sure to point out they are optional.

Author: I will start a new PR for documenting the variables I know, it's not only DEV_MODE

file ``factory-config.yml``.

> It can set in specific branches used for testing and debugging

Reviewer 2: ```suggestion
It can be set in specific branches used for testing and debugging
```

and the source code can be controlled by conditional appends,

> for example, if trying to enable systemd coredump::

Reviewer: ```
The `DEV_MODE` param can be set using the ``ref_options`` stanza in :ref:`factory-config.yml <def-lmp>`
for specific branches used for testing and debugging.  The source code can then be controlled by conditional appends.
For example, if trying to enable systemd coredump::
```

Author: I spent hours yesterday trying to remember the word "stanza". Thanks

  $ cat  meta-subscriber-overrides/meta-subrecipes-core/systemd/systemd_%.bbappend
  PACKAGECONFIG += "${@bb.utils.contains('DEV_MODE', '1', 'coredump', '', d)}"

Another example is to enable the Yocto Project ``IMAGE_FEATURES`` to include

> some development and debug artifacts to the final image,

Reviewer: ```
some development and debug artifacts in the final image,
```

as defined by `Yocto Project image features`_,
the following line can be added to the file
``meta-subscriber-overrides/recipes-samples/images/lmp-factory-image.bb``::

>     IMAGE_FEATURES += "${@bb.utils.contains('DEV_MODE', '1', '', 'debug-tweaks, tools-sdk', d)}"

Reviewer: I don't think a comma is needed in-between `'debug-tweaks, tools-sdk'`?

.. important::

    * The default LmP configuration is to have ``DEV_MODE`` disabled.

    * The default FoundriesFactory configuration is to **not** define
      a development mode.

    * The development mode should be defined to archive the intended goal.

> 

Reviewer 2: Let's include a note here with something like:
DEV_MODE is a helper variable used in LmP to facilitate conditionally enabling debug packages to the image. Additional packages enabled in this mode should be fully supported by the customer.

.. _ref-dev-mode:

Development Mode
=================================

> LmP provides a variable ``DEV_MODE`` to be used to enable a development mode

Reviewer 3: ```suggestion
LmP provides the variable ``DEV_MODE`` that enables a development mode
```
